File: vi/codes/swift/chapter_divide_and_conquer/build_tree.swift
/**
 * File: build_tree.swift
 * Created Time: 2023-09-02
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* Xây dựng cây nhị phân: chia để trị */
func dfs(preorder: [Int], inorderMap: [Int: Int], i: Int, l: Int, r: Int) -> TreeNode? {
    // Dừng khi khoảng của cây con rỗng
    if r - l < 0 {
        return nil
    }
    // Khởi tạo nút gốc
    let root = TreeNode(x: preorder[i])
    // Tra cứu m để chia cây con trái và phải
    let m = inorderMap[preorder[i]]!
    // Bài toán con: xây dựng cây con trái
    root.left = dfs(preorder: preorder, inorderMap: inorderMap, i: i + 1, l: l, r: m - 1)
    // Bài toán con: xây dựng cây con phải
    root.right = dfs(preorder: preorder, inorderMap: inorderMap, i: i + 1 + m - l, l: m + 1, r: r)
    // Trả về nút gốc
    return root
}

/* Xây dựng cây nhị phân */
func buildTree(preorder: [Int], inorder: [Int]) -> TreeNode? {
    // Khởi tạo bảng băm, lưu ánh xạ từ phần tử của duyệt giữa sang chỉ số
    let inorderMap = inorder.enumerated().reduce(into: [:]) { $0[$1.element] = $1.offset }
    return dfs(preorder: preorder, inorderMap: inorderMap, i: inorder.startIndex, l: inorder.startIndex, r: inorder.endIndex - 1)
}

@main
enum BuildTree {
    /* Mã điều khiển */
    static func main() {
        let preorder = [3, 9, 2, 1, 7]
        let inorder = [9, 3, 1, 2, 7]
        print("Pre-order traversal = \(preorder)")
        print("In-order traversal = \(inorder)")

        let root = buildTree(preorder: preorder, inorder: inorder)
        print("The constructed binary tree is:")
        PrintUtil.printTree(root: root)
    }
}

File: vi/codes/java/chapter_divide_and_conquer/build_tree.java
/**
 * File: build_tree.java
 * Created Time: 2023-07-17
 * Author: krahets (krahets@163.com)
 */

package chapter_divide_and_conquer;

import utils.*;
import java.util.*;

public class build_tree {
    /* Xây dựng cây nhị phân: chia để trị */
    static TreeNode dfs(int[] preorder, Map<Integer, Integer> inorderMap, int i, int l, int r) {
        // Kết thúc khi khoảng của cây con là rỗng
        if (r - l < 0)
            return null;
        // Khởi tạo nút gốc
        TreeNode root = new TreeNode(preorder[i]);
        // Truy vấn m để chia cây con trái và phải
        int m = inorderMap.get(preorder[i]);
        // Bài toán con: xây dựng cây con trái
        root.left = dfs(preorder, inorderMap, i + 1, l, m - 1);
        // Bài toán con: xây dựng cây con phải
        root.right = dfs(preorder, inorderMap, i + 1 + m - l, m + 1, r);
        // Trả về nút gốc
        return root;
    }

    /* Xây dựng cây nhị phân */
    static TreeNode buildTree(int[] preorder, int[] inorder) {
        // Khởi tạo bảng băm, lưu trữ ánh xạ từ phần tử duyệt giữa sang chỉ số
        Map<Integer, Integer> inorderMap = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inorderMap.put(inorder[i], i);
        }
        TreeNode root = dfs(preorder, inorderMap, 0, 0, inorder.length - 1);
        return root;
    }

    public static void main(String[] args) {
        int[] preorder = { 3, 9, 2, 1, 7 };
        int[] inorder = { 9, 3, 1, 2, 7 };
        System.out.println("Duyệt trước = " + Arrays.toString(preorder));
        System.out.println("Duyệt giữa = " + Arrays.toString(inorder));

        TreeNode root = buildTree(preorder, inorder);
        System.out.println("Cây nhị phân được xây dựng là:");
        PrintUtil.printTree(root);
    }
}
